internal/metrics/metrics.go

```diff
@@ -27,40 +40,61 @@ const (
 // desired endpoint. It should be invoked on application shutdown.
 type ReportFunc func()
 
-// MeterProvider returns a MeterProvider suitable for CLI usage.
+// Initialize returns a context.Context with a MeterProvider suitable for CLI usage.
 // The primary difference between this metric reader and a more typical
 // usage is that metric reporting only happens once when ReportFunc
 // is invoked.
-func MeterProvider(cli command.Cli) (metric.MeterProvider, ReportFunc, error) {
+func Initialize(ctx context.Context, cli command.Cli) (context.Context, ReportFunc, error) {
 	var exps []sdkmetric.Exporter
 
-	if exp, err := dockerOtelExporter(cli); err != nil {
-		return nil, nil, err
-	} else if exp != nil {
-		exps = append(exps, exp)
-	}
+	// Only metric exporters if the experimental flag is set.
+	if env.IsExperimental() {
+		if exp, err := dockerOtelExporter(cli); err != nil {
+			return nil, nil, err
+		} else if exp != nil {
+			exps = append(exps, exp)
+		}
 
-	if exp, err := detectOtlpExporter(context.Background()); err != nil {
-		return nil, nil, err
-	} else if exp != nil {
-		exps = append(exps, exp)
+		if exp, err := detectOtlpExporter(context.Background()); err != nil {
+			return nil, nil, err
+		} else if exp != nil {
+			exps = append(exps, exp)
+		}
 	}
 
 	if len(exps) == 0 {
 		// No exporters are configured so use a noop provider.
-		return noop.NewMeterProvider(), func() {}, nil
+		return ctx, func() {}, nil
 	}
 
-	// Use delta temporality because, since this is a CLI program, we can never
-	// know the cumulative value.
 	reader := sdkmetric.NewManualReader(
 		sdkmetric.WithTemporalitySelector(deltaTemporality),
 	)
 	mp := sdkmetric.NewMeterProvider(
-		sdkmetric.WithResource(detect.Resource()),
+		sdkmetric.WithResource(Resource()),
```

Comment:
Not new code but why is this duplicated from buildkit?

Reply:
I needed to add new code to include the service id in the resource but didn't want to modify buildkit for that, but I wanted to reuse the logic for the default service ID so I duplicated the code so the two were no longer linked when they had different outputs.